File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccessTesting/FakeVaultSyncPort.swift
import Foundation
import SecretaryVaultAccess

/// In-memory `VaultSyncPort` returning pre-seeded results and spying on inputs.
/// Mirrors `FakeVaultOpenPort`'s convention.
public final class FakeVaultSyncPort: VaultSyncPort {
    private let statusResult: Result<SyncStatus, VaultSyncError>
    private let syncResult: Result<SyncOutcome, VaultSyncError>
    private let commitResult: Result<SyncOutcome, VaultSyncError>

    /// Spies for assertions.
    public private(set) var statusCalls = 0
    public private(set) var syncCalls = 0
    public private(set) var commitCalls = 0
    public private(set) var lastSyncPassword: [UInt8]?
    public private(set) var lastCommitPassword: [UInt8]?
    public private(set) var lastCommitDecisions: [SyncVetoDecision]?
    public private(set) var lastCommitManifestHash: [UInt8]?

    /// Optional rendezvous so an off-main-actor test can hold a call mid-flight.
    public var gate: SuspensionGate?

    public init(statusResult: Result<SyncStatus, VaultSyncError> = .success(
                    SyncStatus(hasState: false, deviceClocks: [], lastStateWriteMs: nil)),
                syncResult: Result<SyncOutcome, VaultSyncError> = .success(.nothingToDo),
                commitResult: Result<SyncOutcome, VaultSyncError> = .success(.mergedClean)) {
        self.statusResult = statusResult
        self.syncResult = syncResult
        self.commitResult = commitResult
    }

    public func status(stateDir: String, vaultUuid: [UInt8]) async throws -> SyncStatus {
        statusCalls += 1
        return try statusResult.get()
    }

    public func sync(stateDir: String, vaultFolder: String,
                     password: [UInt8], nowMs: UInt64) async throws -> SyncOutcome {
        syncCalls += 1
        lastSyncPassword = password
        await gate?.enterAndWait()
        return try syncResult.get()
    }

    public func commitDecisions(stateDir: String, vaultFolder: String,
                                password: [UInt8], decisions: [SyncVetoDecision],
                                manifestHash: [UInt8], nowMs: UInt64) async throws -> SyncOutcome {
        commitCalls += 1
        lastCommitPassword = password
        lastCommitDecisions = decisions
        lastCommitManifestHash = manifestHash
        await gate?.enterAndWait()
        return try commitResult.get()
    }
}

File: android/vault-access/src/test/kotlin/org/secretary/sync/FakeVaultSyncPort.kt
package org.secretary.sync

/**
 * Scriptable in-memory [VaultSyncPort] for host tests. Seed per-method results and inspect
 * the recorded calls (spy). A seeded `Result.failure` is thrown, mirroring how the real
 * adapter surfaces [VaultSyncError].
 *
 * Seeding model: [sync] and [commitDecisions] are FIFO queues (so a multi-pass coordinator
 * test can script a sequence of outcomes); [status] is a single stable value returned on
 * every read.
 */
class FakeVaultSyncPort : VaultSyncPort {
    val syncResults: ArrayDeque<Result<SyncOutcome>> = ArrayDeque()
    val commitResults: ArrayDeque<Result<SyncOutcome>> = ArrayDeque()
    var statusResult: Result<SyncStatus> = Result.success(
        SyncStatus(hasState = false, deviceClocks = emptyList(), lastStateWriteMs = null),
    )

    val syncCalls: MutableList<SyncCall> = mutableListOf()
    val commitCalls: MutableList<CommitCall> = mutableListOf()
    val statusCalls: MutableList<ByteArray> = mutableListOf()

    // Field-access only; never ==-compared, so the ByteArray fields' referential equals/hashCode are unused.
    data class SyncCall(
        val stateDir: String,
        val vaultFolder: String,
        val password: ByteArray,
        val nowMs: ULong,
    )

    // Field-access only; never ==-compared, so the ByteArray fields' referential equals/hashCode are unused.
    data class CommitCall(
        val stateDir: String,
        val vaultFolder: String,
        val password: ByteArray,
        val decisions: List<SyncVetoDecision>,
        val manifestHash: ByteArray,
        val nowMs: ULong,
    )

    override suspend fun status(stateDir: String, vaultUuid: ByteArray): SyncStatus {
        statusCalls += vaultUuid
        return statusResult.getOrThrow()
    }

    override suspend fun sync(
        stateDir: String,
        vaultFolder: String,
        password: ByteArray,
        nowMs: ULong,
    ): SyncOutcome {
        syncCalls += SyncCall(stateDir, vaultFolder, password, nowMs)
        return syncResults.removeFirst().getOrThrow()
    }

    override suspend fun commitDecisions(
        stateDir: String,
        vaultFolder: String,
        password: ByteArray,
        decisions: List<SyncVetoDecision>,
        manifestHash: ByteArray,
        nowMs: ULong,
    ): SyncOutcome {
        commitCalls += CommitCall(stateDir, vaultFolder, password, decisions, manifestHash, nowMs)
        return commitResults.removeFirst().getOrThrow()
    }
}
